### PR TITLE
Day 4. Homework | Day 5. Practice + Homework | Day 6. Homework

### PR DESCRIPTION
**Pull request by eclimov**
**Day 4** 
- Moved question list in store in module question 
- Created api for question EndPoint

**Additionally** 
- Implemented expandable lists for users and tags in left sidebar
- Implemented "store" workflow for tags

**Day 5**
- Implement user list on the right side of the page
- Add users' avatars
- Implement users' search by name
- Add tooltip with some user info for each user avatar

**Day 6**
- Use store in api requests
- Implement view for each object: User(s), Tag(s), Question(s)
- Implement the main search input as autocompleter with groups: it should search by all objects

### CODE REVIEW

File: src/components/content/Questions.vue [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  <div>
+    <div v-for="question in questions" :key="question.id">
+      <div v-html="question.body" />
+      <hr>
+    </div>
+  </div>
+</template>
+
+<script>
+  import {mapActions, mapGetters} from 'vuex'
```
Comment (gregurco) [inline src/components/content/Questions.vue]: please verify code style:
```
import { mapActions, mapGetters } from 'vuex'
```

File: src/components/content/Questions.vue [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  export default {
```
Comment (gregurco) [inline src/components/content/Questions.vue]: all inside this block should be on another level. In your case `computed` is on the same level with `export`, but should not be.

File: src/components/layouts/CommonLayout.vue [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+</template>
+
+<script>
+  import Header from '@/components/layouts/Header'
+  import LeftMenu from '@/components/layouts/LeftMenu'
+  import RightMenu from '@/components/layouts/RightMenu'
+
+  export default {
+  name: 'CommonLayout',
+  components: {RightMenu, LeftMenu, Header }
```
Comment (gregurco) [inline src/components/layouts/CommonLayout.vue]: also code style

File: src/components/content/Questions.vue (hunk 2 above)
Comment (gregurco) [inline src/components/content/Questions.vue]: add name for component

File: src/components/layouts/RightMenu.vue [hunk 4]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (gregurco) [inline src/components/layouts/RightMenu.vue]: remove this extra empty line

File: src/components/rightSidebar/Users.vue [hunk 5]
```diff
@@ -0,0 +1,79 @@
+<template xmlns:v-slot="http://www.w3.org/1999/XSL/Transform">
```
Comment (gregurco) [inline src/components/rightSidebar/Users.vue]: why do you need `xmlns:v-slot="http://www.w3.org/1999/XSL/Transform"`?

File: src/components/rightSidebar/Users.vue [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  <div>
+    <v-text-field
+      label="Search"
+      single-line
+      v-model="userNameLike"
+      :value="userNameLike"
+    ></v-text-field>
```
Comment (gregurco) [inline src/components/rightSidebar/Users.vue]: you can write `<foo />` instead of `<foo></foo>`. Also for `v-progress-circular`.

**Review by gregurco — CHANGES_REQUESTED** [review]

File: src/components/rightSidebar/Users.vue (hunk 5 above)
Comment (eclimov) [reply src/components/rightSidebar/Users.vue]: When I use `v-slot` directive, WebStorm highly recommends generating a namespace declaration. 
Reason: _Namespace 'v-slot' is not bound_

File: src/components/rightSidebar/Users.vue (hunk 6 above)
Comment (eclimov) [reply src/components/rightSidebar/Users.vue]: Didn't quite understand this issue. Can you explain, please?

File: src/components/rightSidebar/Users.vue (hunk 5 above)
Comment (gregurco) [reply src/components/rightSidebar/Users.vue]: I guess it's I bug in webstorm: https://youtrack.jetbrains.com/issue/WEB-37732

File: src/components/rightSidebar/Users.vue (hunk 6 above)
Comment (gregurco) [reply src/components/rightSidebar/Users.vue]: I edited initial comment. Please check it once again. It was bad formatted.

Comment (eclimov) [reply src/components/rightSidebar/Users.vue]: Fixed.
(WebStorm advised this, because `Empty tag doesn't work in some browsers`)

Comment (gregurco) [reply src/components/rightSidebar/Users.vue]: Yes, but this rule is not applicable to frameworks. You can disable "Empty tag doesn't work in some browsers" warning in config.

File: src/components/content/Questions.vue (hunk 1 above)
Comment (gregurco) [reply src/components/content/Questions.vue]: Could you check why this rule is not applies?
Link: https://eslint.org/docs/rules/key-spacing

Comment (eclimov) [reply src/components/content/Questions.vue]: Fixed. 
Looks like it is caused by "Optimize imports" feature.

**Review by gregurco — APPROVED** [review]

File: public/index.html [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
-  </body>
+<head>
+  <meta charset="utf-8">
+  <meta http-equiv="X-UA-Compatible" content="IE=edge">
+  <meta name="viewport" content="width=device-width,initial-scale=1.0">
+  <link rel="icon" href="<%= BASE_URL %>favicon.ico">
+  <title>StackOverflow</title>
+  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900">
+  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Material+Icons">
+    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900">
```
Comment (rapi) [inline public/index.html]: dublicate fonts

File: src/api/comments.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+import API from './api'
+
+export const fetchComments = (param) => API.get('/comments?_limit=3')
```
Comment (rapi) [inline src/api/comments.js]: send params via axios(url,{params})
limit must be dynamic

File: src/assets/logo.svg [hunk 9]
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 87.5 100"><defs><style>.cls-1{fill:#1697f6;}.cls-2{fill:#7bc6ff;}.cls-3{fill:#1867c0;}.cls-4{fill:#aeddff;}</style></defs><title>Artboard 46</title><polyline class="cls-1" points="43.75 0 23.31 0 43.75 48.32"/><polygon class="cls-2" points="43.75 62.5 43.75 100 0 14.58 22.92 14.58 43.75 62.5"/><polyline class="cls-3" points="43.75 0 64.19 0 43.75 48.32"/><polygon class="cls-4" points="64.58 14.58 87.5 14.58 43.75 100 43.75 62.5 64.58 14.58"/></svg>
```
Comment (rapi) [inline src/assets/logo.svg]: put it in public

File: src/components/User/LastUserWidget.vue [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  <div>
+    <p class="title text-lg-center">
+      Users
+    </p>
+    <v-text-field
+      label="Search"
+      @input="(q)=>fetchUsers({q})"
+    />
+    <div v-if="loading" class="text-xs-center mt-5">
+      <v-progress-circular
```
Comment (rapi) [inline src/components/User/LastUserWidget.vue]: put progress in separate component

File: src/components/User/LastUserWidget.vue [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  name: 'LastUserWidget',
+  components: { Preview },
+  computed: {
+    ...mapGetters({
+      userList: 'user/getList',
+      loading: 'user/getLoading'
+    })
+  },
+  created () {
+    this.fetchUsers({ _limit: 51 })
```
Comment (rapi) [inline src/components/User/LastUserWidget.vue:37]: you send it to api but in the end is not used

File: src/components/User/Preview.vue [hunk 12]
```diff
@@ -0,0 +1,38 @@
+<template>
+  <v-tooltip top>
+    <template v-slot:activator="{ on }">
```
Comment (rapi) [inline src/components/User/Preview.vue:3]: Bravo for using slots use it more

File: src/views/Users.vue [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+      <v-progress-circular
+        :width="1"
+        color="red"
+        :size="70"
+        indeterminate
+      />
+    </div>
+
+    <div v-for="user in userList" v-else :key="user.id" style="display: inline-block">
+      <Preview :user="user" />
```
Comment (rapi) [inline src/views/Users.vue]: good that you are using props, but dont send all object, send just id, and take rest of information from store

File: src/views/Users.vue [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    <div v-if="loading" class="text-xs-center mt-5">
+      <v-progress-circular
+        :width="1"
+        color="red"
+        :size="70"
+        indeterminate
+      />
+    </div>
+
+    <div v-for="user in userList" v-else :key="user.id" style="display: inline-block">
```
Comment (rapi) [inline src/views/Users.vue]: dont use v-else and v-for in same element

File: src/store/modules/tag.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+      state.list = tags
+    },
+    mutateLoading: (state, loading) => {
+      state.loading = loading
+    }
+  },
+  getters: {
+    getList: (state) => state.list,
+    getLoading: (state) => state.loading,
+    getByName: (state) => (name) => state.list.find(element => element.name === name)
```
Comment (rapi) [inline src/store/modules/tag.js:28]: user can have same name, dont track it by name

File: src/components/layouts/Header.vue [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+        .finally(() => (this.isLoading = false))
+
+      // Lazily load input items
+      fetch('http://localhost:3001/tags?q=' + val)
+        .then(res => res.json())
+        .then(res => {
+          this.entriesTags = res
+        })
+        .catch(err => {
+          console.log(err)
```
Comment (rapi) [inline src/components/layouts/Header.vue]: dont leave console log

File: src/components/layouts/Header.vue [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+      // Items have already been loaded
+      if (this.items.length > 0) return
+
+      // Items have already been requested
+      if (this.isLoading) return
+
+      this.isLoading = true
+
+      // Lazily load input items
+      fetch('http://localhost:3001/questions?q=' + val)
```
Comment (rapi) [inline src/components/layouts/Header.vue]: bad way, use the function from api

**Review by rapi — CHANGES_REQUESTED** [review]
Setup your eslint better, and look in terminal, please

File: src/store/modules/tag.js (hunk 15 above)
Comment (eclimov) [reply src/store/modules/tag.js:28]: This applies to tags only.
Tags have unique names.

**Comment by eclimov** [comment]
Done

**Pull request merged**